File: src/c#/SimulationObjects/Person.cs
/*
 * created by: b farooq, poly montreal
 * on: 22 october, 2013
 * last edited by: b farooq, poly montreal
 * on: 22 october, 2013
 * summary: 
 * comments:
 */

using System;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using PopulationSynthesis.Utils;

namespace SimulationObjects
{
    sealed class Person : SimulationObject
    {
        private static uint idCounter = 0;

        private HouseholdSize2 HouseholdSize;
        public HouseholdSize2 GetHhldSize()
        {
            return HouseholdSize;
        }
        public void SetHhldSize(HouseholdSize2 size)
        {
            HouseholdSize = size;
        }
        private Age Age;
        public Age GetAge()
        {
            return Age;
        }
        public void SetAge(Age curAge)
        {
            Age = curAge;
        }
        private Sex Sex;
        public Sex GetSex()
        {
            return Sex;
        }
        public void SetSex(Sex curSex)
        {
            Sex = curSex;
        }
        private EducationLevel EducationLevel;
        public EducationLevel GetEducationLevel()
        {
            return EducationLevel;
        }
        public void SetEducationLevel(EducationLevel eduLvl)
        {
            EducationLevel = eduLvl;
        }

        private string ZoneID;
        public string GetZoneID()
        {
            return ZoneID;
        }
        public void SetZoneID(string id)
        {
            ZoneID = id;
        }

        public Person()
        {
            Sex = Sex.Female;
            Age = Age.ThirtyFiveToFortyFour;
            EducationLevel = EducationLevel.primary;
            HouseholdSize = HouseholdSize2.TwoPersons;
            Type = AgentType.Person;
        }

        public Person(string currZone)
        {
            ZoneID = currZone;
            Sex = Sex.Female;
            Age = Age.ThirtyFiveToFortyFour;
            EducationLevel = EducationLevel.primary;
            HouseholdSize = HouseholdSize2.TwoPersons;
            Type = AgentType.Person;
        }

        private Person(Person original)
        {
            Type = AgentType.Person;
            //copy the values
            ZoneID = original.ZoneID;
            Age = original.Age;
            Sex = original.Sex;
            HouseholdSize = original.HouseholdSize;
            EducationLevel = original.EducationLevel;
        }

        [ThreadStatic]
        private static StringBuilder KeyBuilder;

        public override string GetNewJointKey(string baseDim)
        {
            var builder = KeyBuilder;
            if(builder == null)
            {
                KeyBuilder = builder = new StringBuilder();
            }
            builder.Clear();
            if(baseDim == "Age")
            {
                builder.Append((int)Sex);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)HouseholdSize);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)EducationLevel);
            }
            else if(baseDim == "Sex")
            {
                builder.Append((int)Age).ToString();
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)HouseholdSize);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)EducationLevel);
            }
            else if(baseDim == "HouseholdSize2")
            {
                builder.Append((int)Age);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)Sex);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)EducationLevel);
            }
            else if(baseDim == "EducationLevel")
            {
                builder.Append((int)Age);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)Sex);
                builder.Append(Constants.CONDITIONAL_DELIMITER);
                builder.Append((int)HouseholdSize);
            }
            return builder.ToString();
        }

        public override SimulationObject CreateNewCopy(string baseDim,
            int baseDimVal)
        {
            Person copy = new Person(this);
            // apply the new change
            switch(baseDim)
            {
                case "Age":
                    copy.Age = (Age)baseDimVal;
                    break;
                case "Sex":
                    copy.Sex = (Sex)baseDimVal;
                    break;
                case "HouseholdSize2":
                    copy.HouseholdSize = (HouseholdSize2)baseDimVal;
                    break;
                case "EducationLevel":
                    copy.EducationLevel = (EducationLevel)baseDimVal;
                    break;
                default:
                    return null;
            }
            return copy;
        }
    }
}

File: src/SimulationObjects/Person.java
package SimulationObjects;

import SimulationObjects.AgentAttributes.*;
import Utils.Utils;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.logging.Level;
import java.util.logging.Logger;

/*
 * created by: b farooq, poly montreal
 * on: 22 october, 2013
 * last edited by: b farooq, poly montreal
 * on: 22 october, 2013
 * summary: 
 * comments:
 */

public class Person extends SimulationObject implements Serializable
{
    private static int idCounter = 0;

        private int myHhldSize;
        public int GetHhldSize()
        {
            return myHhldSize;
        }
        public void SetHhldSize(int size)
        {
            myHhldSize = size;
        }
        private int myAge;
        public int GetAge()
        {
            return myAge;
        }
        public void SetAge(int curAge)
        {
            myAge = curAge;
        }
        private int mySex;
        public int GetSex()
        {
            return mySex;
        }
        public void SetSex(int curSex)
        {
            mySex = curSex;
        }
        private int myEducLevel;
        public int GetEducationLevel()
        {
            return myEducLevel;
        }
        public void SetEducationLevel(int eduLvl)
        {
            myEducLevel = eduLvl;
        }

        private String myZoneID;
        public String GetZoneID()
        {
            return myZoneID;
        }
        public void SetZoneID(String id)
        {
            myZoneID = id;
        }

        public Person()
        {
            myID = ++idCounter;
            mySex = Sex.Female;
            myAge = Age.ThirtyFiveToFortyFour;
            myEducLevel = EducationLevel.primary;
            myHhldSize = HouseholdSize2.TwoPersons;
            myType = AgentType.Person;
        }

        public Person(String currZone)
        {
            myZoneID = currZone;
            myID = ++idCounter;
            mySex = Sex.Female;
            myAge = Age.ThirtyFiveToFortyFour;
            myEducLevel = EducationLevel.primary;
            myHhldSize = HouseholdSize2.TwoPersons;
            myType = AgentType.Person;
        }

    @Override
        public String GetNewJointKey(String baseDim)
        {
            String jointKey;
            if (baseDim.equals("Age"))
            {
                jointKey = String.valueOf(mySex)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myHhldSize)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myEducLevel);
            }
            else if (baseDim.equals("Sex"))
            {
                jointKey = String.valueOf(myAge)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myHhldSize)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myEducLevel);
            }
            else if (baseDim.equals("HouseholdSize2"))
            {
                jointKey = String.valueOf(myAge)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(mySex)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myEducLevel);
            }
            else if (baseDim.equals("EducationLevel"))
            {
                jointKey = String.valueOf(myAge)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(mySex)
                            + Utils.CONDITIONAL_DELIMITER
                            + String.valueOf(myHhldSize);
            }
            else
            {
                jointKey = "";
            }

            return jointKey;
        }

    @Override
        public SimulationObject CreateNewCopy(String baseDim, int baseDimVal)
        {
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            ObjectOutputStream oos;
            try {
                oos = new ObjectOutputStream(bos);
                oos.writeObject(this);
                oos.flush();
                oos.close();
                bos.close();
                byte [] byteData = bos.toByteArray();

                ByteArrayInputStream bais = new ByteArrayInputStream(byteData);
                try
                {
                    Person myCopy = (Person)new ObjectInputStream(bais).readObject();

                    if (baseDim.equals("Age"))
                    {
                        myCopy.myAge = baseDimVal;
                    }
                    else if (baseDim.equals("Sex"))
                    {
                        myCopy.mySex = baseDimVal;
                    }
                    else if (baseDim.equals("HouseholdSize2"))
                    {
                        myCopy.myHhldSize = baseDimVal;
                    }
                    else if (baseDim.equals("EducationLevel"))
                    {
                        myCopy.myEducLevel = baseDimVal;
                    }
                    else
                    {
                        return null;
                    }
                    return myCopy;

                } catch (ClassNotFoundException ex) {
                    Logger.getLogger(Household.class.getName()).log(Level.SEVERE, null, ex);
                }

            } catch (IOException ex) {
                Logger.getLogger(Household.class.getName()).log(Level.SEVERE, null, ex);
            }
            return null;
        }
}
